src/Marten/Schema/NgramIndex.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using Baseline;
+using Weasel.Core;
+using Weasel.Postgresql;
+using Weasel.Postgresql.Tables;
+
+namespace Marten.Schema
+{
+    /// <summary>
+    /// Implements an index that extracts ngrams for the specified field.
+    /// </summary>
+    public class NgramIndex: IndexDefinition
```

Comment:
**CONCERN:** @apereiratl, do we have a specific reason why we cannot reuse `FullTextIndex` for NGram?

Reply:
To be honest, when I initially tested Marten, I don't believe this worked for me. I will retest this and update the comment.

Reply:
Created a little script to test it once again. 
```
void Main()
{
	var store = GetConfiguredStore();
	using var session = store.OpenSession();
	// session
	
	var i = 0;
	while(i < 100)
	{
		var user = new UserTest4x
		{
			FirstName = System.Guid.NewGuid().ToString("B"),
			LastName = System.Guid.NewGuid().ToString("B"),
		};
		
		session.Store(user);
		i++;
	}
	
	session.SaveChanges();
	
	var randomUser = session.Query<UserTest4x>().OrderBy(ut => ut.FirstName).First();
	randomUser.Dump();
	
	IQueryable<UserTest4x> users = session.Query<UserTest4x>().Stats(out var stats);
	
	users = users.Where(x => x.FirstName.NgramSearch(randomUser.FirstName.Substring(0, 3)));
	
	var matchedUsersNgram = users.ToList().Dump();

	IQueryable<UserTest4x> usersx = session.Query<UserTest4x>().Stats(out var statsx);

	usersx = usersx.Where(x => x.FirstName.PlainTextSearch(randomUser.FirstName.Substring(0, 3)));

	
	var matchedUsersFullText = usersx.ToList().Dump();
	
	matchedUsersNgram.Except(matchedUsersFullText).ToList().Dump();

	$"Count from Ngram {matchedUsersNgram.Count()} Count from Full text search {matchedUsersFullText.Count()}".Dump();
}

/// <summary>
/// Returns a configured data store.
/// </summary>
/// <param name="settings">
/// The settings for the store.
/// </param>
/// <returns>
/// The <see cref="IDocumentStore"/>.
/// </returns>
public static IDocumentStore GetConfiguredStore()
{
	var store = DocumentStore.For(_ =>
   {
	   _.AutoCreateSchemaObjects = AutoCreate.CreateOrUpdate;
	   _.Advanced.HiloSequenceDefaults.MaxLo = 30;
	   _.Connection("host=localhost;database=some_db;password=some_db_pass;username=some_db_usr;Enlist=false");
	   _.Schema.Include<MyMartenRegistry>();
	   _.UseNodaTime();
	   _.Logger(new ConsoleMartenLogger());
   });

	return store;
}

public class MyMartenRegistry : MartenRegistry
{
	public MyMartenRegistry()
	{
		this.For<UserTest4x>()
				.Duplicate(x => x.FirstName)
				.NgramIndex(x => x.FirstName)

				.Duplicate(x => x.LastName)
				.NgramIndex(x => x.LastName)
				
				.FullTextIndex(x => x.FirstName)
				
				.GinIndexJsonData();
	}
}

public class UserTest4x
{
	public long Id { get; set; }
	
	public string FirstName { get; set; }
	
	public string LastName { get; set; }
}
```

Here's the final output.

```
select d.id, d.data from public.mt_doc_userquery_usertest4x as d order by d.first_name LIMIT :p0
  p0: 1
select d.id, d.data, count(1) OVER() as total_rows from public.mt_doc_userquery_usertest4x as d where mt_grams_vector(d.first_name) @@ mt_grams_query(:p0)
  p0: {00
select d.id, d.data, count(1) OVER() as total_rows from public.mt_doc_userquery_usertest4x as d where to_tsvector(:argRegConfig::regconfig, ((first_name))) @@ plainto_tsquery(:argRegConfig::regconfig, :argSearchTerm)
  argRegConfig: english
  argSearchTerm: {00
Count from Ngram 82 Count from Full text search 0
```